### PR TITLE
Update database readonly mode

### PR DESCRIPTION
**Pull request by calvinlu3**
This fixes [issue 2838](https://github.com/oncokb/oncokb/issues/2838)

- Created a `ReadOnlyMode` HOC that will displays a generic unavailable page
- Repository save() calls will throw an `DatabaseReadOnlyException` with status code 503 (Service Unavailable) when trying to write

### CODE REVIEW

**Comment by calvinlu3** [comment]
- [ ] Why use Transactional annotation and advantages?
- [ ] Identify places where Transactional is use and the object is modified

**Comment by calvinlu3** [comment]
Reviewing the failed tests in `ExceptionTranslatorIT.java`

**Comment by calvinlu3** [comment]
Okay, seems to have to do with the ControllerAdvice I defined. Will try to move the exception handling into ExceptionTranslator.java instead.

File: src/main/webapp/app/shared/readonly/ReadOnlyMode.tsx [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import { Alert } from 'react-bootstrap';
+import ReadOnlyPage from './ReadOnlyPage';
+
+// Todo:
+// We now have multiple types of routes (RecaptchaBoundaryRoute, PrivateRoute, and now a read only route)
+// We can combine these routes into a general route component and just specify the route type(s).
+
+/**
+ * Components wrapped in the HOC will display a warning page or have a warning message added to the page
+ * when application is is read only mode.
```
Comment (zhx828) [inline src/main/webapp/app/shared/readonly/ReadOnlyMode.tsx:12]:
```suggestion
 * when application is in read only mode.
```

File: src/main/webapp/app/shared/readonly/ReadOnlyMode.tsx [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * @param Component the original component
+ * @param showWarningOnly if true and app is in read only mode, then the original component
+ * will be displayed and a warning message will be added to the top of the page.
+ */
+
+export default function ReadOnlyMode(
+  Component: any,
```
Comment (zhx828) [inline src/main/webapp/app/shared/readonly/ReadOnlyMode.tsx:19]: I think the type should at least be a JSX.Element so you can render?

File: src/main/webapp/app/shared/readonly/ReadOnlyMode.tsx [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+) {
+  return (props: any) => {
+    if (AppConfig.serverConfig.readonly) {
+      if (showWarningOnly) {
+        return (
+          <>
+            <Alert variant="danger">
+              <h4>This page is read-only</h4>
+              <div>
+                Any changes you make won't be saved at this time. The page will
```
Comment (zhx828) [inline src/main/webapp/app/shared/readonly/ReadOnlyMode.tsx:30]: Let's emphasize using `WON'T` or `will NOT`?

File: src/main/webapp/app/shared/readonly/ReadOnlyPage.tsx [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+import { Alert } from 'react-bootstrap';
+import { observer } from 'mobx-react';
+
+const DEFAULT_TITLE = 'This page is temporarily unavailable.';
```
Comment (zhx828) [inline src/main/webapp/app/shared/readonly/ReadOnlyPage.tsx:5]: May not be necessary if it's not been shared?

File: src/main/java/org/mskcc/cbio/oncokb/aop/repository/DatabaseReadOnlyAspect.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+    @Pointcut("execution(* org.mskcc.cbio.oncokb.repository.*.save(..))")
+    private void anyRepositorySaveOperation() {}
+
+    @Before(value = "anyRepositorySaveOperation() && excludedRepositorySaveOperation()")
+    public void databaseReadOnlyWithError(JoinPoint jp) throws Throwable {
+        if(Optional.ofNullable(this.applicationProperties.getDbReadOnly()).orElse(false)){
+            // When the save method is called, we throw an exception that will be 
+            // handled by DatabaseReadOnlyAdvisor (ControllerAdvice).
+            throw new DatabaseReadOnlyException();
```
Comment (zhx828) [inline src/main/java/org/mskcc/cbio/oncokb/aop/repository/DatabaseReadOnlyAspect.java:38]: Can we skip this exception in `initSentry` of `OncokbPublicApp.java`?

File: src/main/java/org/mskcc/cbio/oncokb/web/rest/errors/DatabaseReadOnlyException.java [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+
+import org.zalando.problem.AbstractThrowableProblem;
+import org.zalando.problem.Status;
+
+public class DatabaseReadOnlyException extends AbstractThrowableProblem{
+
+    private static final long serialVersionUID = 1L;
+
+    public DatabaseReadOnlyException() {
+        super(ErrorConstants.SERVICE_UNAVAILABLE, "Service is temporarily unavailable", Status.SERVICE_UNAVAILABLE);
```
Comment (zhx828) [inline src/main/java/org/mskcc/cbio/oncokb/web/rest/errors/DatabaseReadOnlyException.java:11]: Let's be clear here. Maybe `We are not able to update the content at this moment. Service should be back momentarily.`?

File: src/main/java/org/mskcc/cbio/oncokb/web/rest/errors/ErrorConstants.java [hunk 7]
```diff
@@ -14,6 +14,7 @@
     public static final URI LOGIN_ALREADY_USED_TYPE = URI.create(PROBLEM_BASE_URL + "/login-already-used");
     public static final URI TOKEN_EXPIRED_TYPE = URI.create(PROBLEM_BASE_URL + "/token-expired");
     public static final URI LICENSE_AGREEMENT_NOT_ACCEPTED = URI.create(PROBLEM_BASE_URL + "/license-agreement-not-accepted");
+    public static final URI SERVICE_UNAVAILABLE = URI.create(PROBLEM_BASE_URL + "/service-unavailable");
```
Comment (zhx828) [inline src/main/java/org/mskcc/cbio/oncokb/web/rest/errors/ErrorConstants.java:17]: Let's just use DATABASE_READ_ONLY.

**Review by zhx828 — CHANGES_REQUESTED** [review]
Overall looks good. Some minor comments.

**Comment by calvinlu3** [comment]
@zhx828 Thanks for review, will get these adjusted. 👍

**Review by zhx828 — APPROVED** [review]
LGTM 👍🏼

**Pull request merged**